FEC-11577 AdLayout Configuration

GouravSna opened the pull request:
### Description of the Changes

[Works with [Kaltura-Player](https://github.com/kaltura/kaltura-player-android) & [Playkit-IMA](https://github.com/kaltura/playkit-android-ima)] 

With Ad layout config you can create your own ad break timeline using your vast tags. Ad break can be set as pre, mid and post rolls and each ad break can contain a single vast tag or multiple tags, either as a pod, but also as a Waterfall.

GouravSna committed:
- Ad waterfalling

custadian committed:
- Barebone structure for AdLayout

GouravSna committed:
- Fix me

GouravSna committed:
- Added an advertising tree for internal use

GouravSna committed:
Merge branch 'dev' into FEC-11577

GouravSna committed:
- Handling Seek condition

GouravSna committed:
- Fixed in between seeking situation
- Refactored queue to list

GouravSna committed:
- handle waterfalling

GouravSna committed:
- Fixed issue if seek happens between pre-roll and immediate next midroll

GouravSna committed:
Merge branch 'dev' into FEC-11577

GouravSna committed:
- allAdsCompleted event masking

GouravSna committed:
- API updates

GouravSna committed:
- Controller refactor

GouravSna committed:
- Code refactor

GouravSna committed:
- Multiple AdPod support in one AdBreak

GouravSna committed:
- Handling multiple AdPods in AdBreak

GouravSna committed:
Merge branch 'dev' into FEC-11577

GouravSna committed:
- Handling AdPod State in case if all ads are errored

GouravSna committed:
- Frequency based (EVERY) Ad playback

GouravSna committed:
- Play content player if preroll ad is not configured

GouravSna committed:
- Fixed only postroll playback

GouravSna committed:
- Percentage based Ad configuration
- JSON support for Ad Confguration
- Moved internal classes in a single container
- Renamed AdvertisingTree to AdvertisingContainer
- Code refactor and cleanup
- Added Javadocs

GouravSna committed:
- Fixed updation logic in cuepoints

GouravSna committed:
- doc update

GouravSna committed:
- If user seek to the end so play the last unplayed midroll and then postroll

GouravSna committed:
- Added TimeUnit for Ads Config
- Added AdEvent for waterfalling

giladna committed:
Merge branch 'dev' into FEC-11577-gradlev7

GouravSna committed:
- Fixed issue when midroll was not being played if preroll is not given

giladna committed:
Merge branch 'FEC-11577' into FEC-11577-gradlev7

GouravSna committed:
- ChangeMedia functionality
- Config cleanup

GouravSna committed:
- Code cleanup

giladna committed:
Merge branch 'FEC-11577' into FEC-11577-gradlev7

GouravSna committed:
- Made adTime unit default as seconds
- All ads completed (WIP)
- Releasing resources of controlled in destroy()

giladna committed:
Merge branch 'FEC-11577' into FEC-11577-gradlev7

GouravSna committed:
- All_Ads_Completed masking
- Fixing EVERY issue  ( was not multiplying with 1000L)

giladna committed:
Merge branch 'FEC-11577' into FEC-11577-gradlev7

GouravSna committed:
- Fixing EVERY issue

GouravSna committed:
- Masked AllAdsCompleted event
- Play upcoming ad if seeked with EVERY midroll configuration

GouravSna committed:
- Code refactor with PlayAdNow API

GouravSna committed:
- PlayAdNow api

GouravSna committed:
- Fixed PlayerLoader possible crash
- Fixed crash when immediately pressing back button on app launch

GouravSna committed:
- Code rafactor
- Added playAdaftertime api
- fixed video frame during ad playback

giladna committed:
Merge branch 'FEC-11577' into FEC-11577-gradlev7

GouravSna committed:
- playAdNow API cleanup
- playerAdNowAfter time API

GouravSna committed:
- Live Media support (only pre-roll will be played, later config will be dropped)

GouravSna committed:
- CuePoints updation

GouravSna committed:
- AdInfo for IMAPlugin

GouravSna committed:
- AdWaterfalling, AdWaterfallingFailed events support

giladna committed:
Merge branch 'FEC-11577' into FEC-11577-gradlev7

GouravSna committed:
- AdWatefalling support

GouravSna committed:
- Fixed: postroll was always being added in cuepoint list
- Fixed content playback issues
- Fixed crash when there is only pre-roll
- Support for cuepoint event for EVERY midroll ad type

GouravSna committed:
- Code refactor

GouravSna committed:
- Removed unused code
- PlayAdNow with live media enabled

giladna commented on playkit/src/main/java/com/kaltura/playkit/PlayerDecoratorBase.java:
```diff
@@ -37,6 +39,11 @@ public void prepare(@NonNull PKMediaConfig mediaConfig) {
     }
 
+    @Override
+    public void setAdvertising(@NonNull AdvertisingConfig advertisingConfig, @NonNull PKAdvertisingController pkAdvertisingController) {
```
i think switching the order of params required first controller than config?

GouravSna replied:
Done

giladna commented on playkit/src/main/java/com/kaltura/playkit/plugins/ads/AdEvent.java:
```diff
@@ -253,7 +277,9 @@ public DAISourceSelected(String sourceURL) {
         ERROR,
-        DAI_SOURCE_SELECTED
+        DAI_SOURCE_SELECTED,
+        AD_WATERFALLING,
```
@GouravSna  what is this event meaning

GouravSna replied:
This is when the AdWaterfalling start then it will be fired and if all the ads failed in the waterfalling list then AD_WATERFALLING_FAILED will be fired.

giladna committed:
Update PKAdvertising.kt

giladna committed:
Update IMAEventsListener.kt

giladna committed:
Update AdvertisingContainer.kt

giladna commented on playkit/src/main/java/com/kaltura/playkit/ads/AdvertisingConfig.kt:
```diff
@@ -0,0 +1,85 @@
+
+/**
+ * Advertising configuration
+ * Pass the list the AdBreaks
```
list of?

giladna commented on playkit/src/main/java/com/kaltura/playkit/ads/AdvertisingConfig.kt:
```diff
@@ -0,0 +1,85 @@
+                             @NonNull val playAdsAfterTime: Long = Long.MIN_VALUE)
+
+/**
+ * AdBreak: Pre, Mid, Post
```
type?

giladna replied:
@GouravSna  java docs is missing  @NonNull val adType: AdType = AdType.AD_URL,

giladna commented on playkit/src/main/java/com/kaltura/playkit/ads/AdvertisingConfig.kt:
```diff
@@ -0,0 +1,85 @@
+}
+
+// Adroll type
+internal enum class AdrollType {
```
AdRollType?

giladna replied:
@GouravSna  this was fixed?

giladna committed:
Merge branch 'FEC-11577' into FEC-11577-gradlev7

GouravSna committed:
- Corrected comment

GouravSna committed:
- PlayAdNow support with JSON

GouravSna committed:
- JsonSyntaxException updated

GouravSna committed:
- fixed: playback did not start. If preroll is there but playadsafter time is configured with later value,

GouravSna committed:
- Code refactor and cleanup

GouravSna committed:
- If adTag is not empty then only set the AdInfo

GouravSna committed:
- Removed misleading log

GouravSna committed:
- Protected adIndex

GouravSna committed:
- Fix for FEC-11750

GouravSna committed:
Merge branch 'dev' into FEC-11577

GouravSna committed:
- Changed millisecond multiplier with Constant

GouravSna committed:
- Fixed FEC-11754
- Fire all ads completed if it is postroll and there are still unplayed ads

GouravSna committed:
- Fixed EVERY based midroll's behaviour.
- Code refactor as per the PR comments.

GouravSna committed:
- Fixed magic no 3 for EVERY ads

GouravSna committed:
- Fixed AdPlayhead Updated logic

GouravSna committed:
- FIxed/Tested TODOs

GouravSna committed:
- PlayAdNow fix for LiveMedia if allAdsCompleted is fired then also it will work.

GouravSna committed:
- Fixed FEC-11776
- Don't change the original advertising config object

GouravSna committed:
- Fixed FEC-11749

GouravSna committed:
- Fixed FEC-11772 and FEC-11773

GouravSna committed:
- Fixed FEC-11790

GouravSna committed:
- Fixed FEC-11751

GouravSna committed:
- PlayAdsAfterTime while seeking for EVERY

GouravSna committed:
- Fixed FEC-11752 (Pausing the player before making ad request)

GouravSna committed:
- Fixed FEC-11796

GouravSna committed:
- Fixed FEC-11769

GouravSna committed:
- Removed redundant import

GouravSna committed:
- Removed unnecessary values in AdInfo

GouravSna committed:
- Method name refactor

GouravSna committed:
- Added comments

GouravSna committed:
- Fixed Start Position Issue

- Code refactor, Outside world exposed code moved internally on AdsProvider (Removed from AdController)

GouravSna committed:
- Added JavaDoc

GouravSna committed:
- Added JavaDoc

GouravSna committed:
- Code refactor and cleanup

GouravSna committed:
- Code cleanup

GouravSna committed:
Merge branch 'dev' into FEC-11577
